src/TreeContainer.js

```diff
@@ -78,7 +78,7 @@ function render(component) {
         ...children
     );
 
-    return <NotifyObservers id={componentProps.id}>{parent}</NotifyObservers>;
+    return <NotifyObservers key={componentProps.id} id={componentProps.id}>{parent}</NotifyObservers>;
```

Comment:
What if components doesn't have an id ?

Reply:
Well, then the component won't have a `key` either. In the case of the `Graph` component there's always an id - because if there isn't one defined it generates a random one. We could generate something here if id isn't set, or we could make sure that all components that are having trouble with shared state always have an id. Not sure what makes more sense.

Reply:
I had tried to make a key generator with the position in the array of children and type or the id if available. The key looked like:

`Tabs/0/Tab/#output`

I don't know if I still have that code, it was in the component as props PR but it didn't solve what I wanted so I may have stripped it out.

Reply:
Could we be more explicit about only providing the key if `id` is defined? Right now, we'll be setting it to `undefined` which might work, but I think it'd be preferrable to just not set it if it's not defined. Otherwise, we'll likely get warnings like "Key is undefined which isn't valid" (https://github.com/facebook/react/issues/11935)

Reply:
I think React won't add the key prop if it's assignment is undefined, but now that you mention it I'm not sure. Will check this out.

Reply:
Feel free to dig in, but this would be a very simple fix that would also improve readability. Otherwise, folks will always wonder "What happens if this is undefined?" when they read the code.
```
extra_props = {}
if (R.has('id', props)) {
   extra_props.key = props.id;
}
return <NotifyObservers {...extra_props}/>
```